src/SegmentedString.chpl

```diff
@@ -1301,7 +1301,7 @@ module SegmentedString {
       const rh = rss.siphash();
       return if polarity then (lh == rh) else (lh != rh);
     }
-    ref oD = lss.offsets.aD;
+    ref oD = lss.offsets.a.domain;
```

Comment:
should we make this a `const ref` as well?

Reply:
Yup, good catch.`array.domain` returns an immutable thing, but using `const ref` allows us to avoid making a copy since `const aD = array.domain` would result in a domain copy, which we want to avoid because copying distributed-domains isn't free).

> To be honest im also not sure when it's appropriate to use const ref for a domain vs just const

My understanding is:

```chpl
    const aD = a.domain; // creates an immutable copy of a.domain
      var aD = a.domain; // creates a mutable copy of a.domain
const ref aD = a.domain; // captures an immutable reference to a.domain
      ref aD = a.domain; // should be illegal since `a.domain` cannot be modified.
```

For arkouda I don't think there's every a reason to prefer `const aD = a.domain;`. The only case I could imagine this being useful is if you have a non-distributed domain and you want to localize it to a new node. Something like:

```chpl
const d = {1..10};
on Locales[1] {
  const localD = d;
}
```

to avoid having to do comm when accessing localD

I'm also surprised the compiler didn't error about this `ref` since that should be illegal. I wonder if this code isn't called or there's a missing const check in the compiler somewhere.

---

And not done in Arkouda, but just for context/background on why `.domain` returns an immutable thing -- the way to resize arrays in chapel is to resize the domain it's declared over. So if you want to grow an array you'd do something like:

```chpl
var d = {1..10};
var A, B: [d] int;
writeln(A.size); // 10
d = {1..20};
writeln(A.size); // 20
```

But we don't want to support resizing through `.domain` for performance reasons and because we'd think it'd be confusing for something like `A.domain = {1..20};` to resize other arrays as well. https://chapel-lang.org/docs/language/spec/arrays.html#association-of-arrays-to-domains has some more info on this.

Reply:
Thanks for the thorough explanation!